### ISSUE TITLE
RFC Expand 2sxc Web API to return ADAM files and folders

### ISSUE THREAD

**Issue by PagnoDunadan** [2sic/2sxc#1640]
**I'm submitting a ...**
[x] feature request

**...about**
[x] APIs like REST

**Current behavior**
Through Web API we can fetch ContentType and Query, but we can't fetch ADAM files. (https://github.com/2sic/2sxc/wiki/webapi)

**Expected behavior**
Web API should return files and folders just like other data from ContentType.

**What is the motivation / use case for changing the behavior?**
Using AsAdam in Razor template to fetch files works fine for small amount of data, but for tables with 1000+ entities it takes 10+ seconds to build the page and we can't load files lazily.

**Comment by iJungleboy** [comment]
To make sure I understand the request properly: you would like to retrieve the list (not the files) as JSON, for use in JS applications - correct?

**Comment by PagnoDunadan** [comment]
For example, [root-path]/app/auto/content/BlogPost/1050 returns data for blog post 1050, but if that post contains Link/file reference field, that field's value will be null in returned JSON:
```
[{
	"Field1":"Value1",
	"Field2":"Value2",
	"FileField":null
},{
	"Field1":"Value3",
	"Field2":"Value4",
	"FileField":null
}]
```
I would like to get those files/folders (objects with Name, URL, etc.).

**Comment by iJungleboy** [comment]
Let's discuss (that's why I renamed it to RFC). What are the most common use cases

1. I want the files in the same request (this could result in an unexpected amount of data)
2. I want to request the files separately (since I know the entity-IDs), so I could request the files for the field "Album" and "Thumbnail" but leave out the large list of "CaseStudies" which may contain 75 documents

More questions
1. would we usually want to know the IDs, or need the URLs?

**Comment by PagnoDunadan** [comment]
I guess that file name, URL and file size are most needed. It can be a separate request to API like ```[root-path]/app/auto/content/BlogPost/1050/Album```. It would return a singular file if there is only one, but multiple files sorted in their respective folders if file library.

**Comment by iJungleboy** [comment]
From what I see, the API is almost there :) - we built it a few months ago to standardize how ADAM works and to enable public forms. 
As of now, it still needs a few params more, and you must set security on that field to get it to work. I'll see if I can document / simplify it in future.

**Comment by PagnoDunadan** [comment]
Awesome! Thank you :)

**Comment by iJungleboy** [comment]
I've opened the API a bit and it will be in RC3. I just don't have time to test it :) so I'll need your help. I'll tell you more about it when we release 9.33 RC3

**Comment by iJungleboy** [comment]
I just released 9.33 and it should actually provide this already - could you give it a try? the format is as you expected it:
`[root-path]/app/auto/content/BlogPost/1050/Album`

Please test, and give me feedback :)

**Comment by PagnoDunadan** [comment]
I will check it out and report back to you, probably no later than Monday. Thank you :)

**Comment by PagnoDunadan** [comment]
### Results using 9.33 Final

Tested both Default and File Library.

### Common issues:

- Fetching with EntityGuid works, but doesn't work with EntityId. **Example:** ![entityid](https://user-images.githubusercontent.com/7113033/48258975-d17f6280-e416-11e8-9664-836cb79535f7.png)

- Created and Url information is missing. **Example:** Adam contains information about when file was Created and it's Url, but API only returns path. Url can be constructed from path adding `/Portals/PortalId/` to the start of the path, but it's not as pretty: ![c](https://user-images.githubusercontent.com/7113033/48259142-63876b00-e417-11e8-82f1-13bd1f970627.png)


### Default specific issues:

- API returns all files ever uploaded, just like File Library, but default field type should only hold latest uploaded file. **Example:** Test field is of type default and I've uploaded 2 files, one after the other, and in 2sxc management menu I can only see the second file, but API returns both files: ![default](https://user-images.githubusercontent.com/7113033/48257244-3b493d80-e412-11e8-8dd8-44825d84ff26.png)


### File Library specific issues:

- API returns only first level files and folders. Looks like there is no way to fetch files from subfolders. **Example:** ImageLibrary field is of type library and contains 10 images and 1 folder TestFolder. TestFolder contains one more image. API returns 10 images and 1 folder, but doesn't return image inside TestFolder: ![library](https://user-images.githubusercontent.com/7113033/48257961-5321c100-e414-11e8-8a7f-764c99810d9b.png) Accessing TestFolder with `[root-path]/app/auto/content/ContentType/EntityGuid/Folder/Subfolder` returns:
![subfolder](https://user-images.githubusercontent.com/7113033/48258167-c4fa0a80-e414-11e8-8866-cad093415eab.png)

**Comment by iJungleboy** [comment]
@PagnoDunadan thanks for this comprehensive testing!

1. guid vs id: At the moment I want to keep it guid only, because whoever creates the JS application should be able to provide guids (unless I'm mistaken, then tell me), and this should just add another layer of security, because for other items the guid will not be known to the public. 
1. path without root - you are probably right that this would be helpful, I'll have to think about what the best implementation is for final, but I'm not sure yet. I can see good reasons to include and exclude, but it's something we won't be able to change once it's final
1. file-field - in this case adam should not be used, as the file field itself already has the value. so I don't think we'll work on that. 
1. subfolders - I believe ATM there's a parameter "subfolder" to browse deeper, but I think I'll rename it to folder in the next release. So you can give it a try, but note that this param will probably be renamed. 

ok?

**Comment by PagnoDunadan** [comment]
I agree that urls for files should probably never be hackable and guid is a better approach. I have tested subfolder parameter and it works well, API returns subfolders and deeper subfolders properly.

Only functionality that prevents me from using API is missing **Created** because I really depend on that information. I have an application where users submit files as a part of the form, but form sometimes takes multiple files and I use Created to find all files submitted in the latest form. I'm assuming that all files created in a one second window belong to the latest iteration of that form. Maybe I could have prepended some id or added some metadata to differentiate files, but the damage is done and now I depend on Created.

**Comment by iJungleboy** [comment]
I made the changes and they won't be out till 9.34. I've attached the modified dlls - pls give it a try. the subfolder parameter is now already called "folder"

[ToSic.SexyContent.WebApi.zip](https://github.com/2sic/2sxc/files/2571529/ToSic.SexyContent.WebApi.zip)

**Comment by PagnoDunadan** [comment]
API now provides Created and Modified and folder parameter works. Thank you for a very fast response :)

**Comment by iJungleboy** [comment]
You're welcome :)

**Comment by PagnoDunadan** [comment]
Setting only Read results in 403. Setting CRUD works.

![read](https://user-images.githubusercontent.com/7113033/48351674-1ad7e800-e68b-11e8-9417-2fa9e099688a.png)

**Comment by iJungleboy** [comment]
Could you check the event log and tell me, what was actually logged? just so I know where it fails. thx

**Comment by iJungleboy** [comment]
...ideally you could also include the full log from the 2sxc-insights, as that has a detailed log why this decision was made. https://2sxc.org/en/blog/post/using-2sxc-insights

**Comment by iJungleboy** [comment]
I think I found it - could you check the new dlls?
[ToSic.WebApi.zip](https://github.com/2sic/2sxc/files/2572299/ToSic.WebApi.zip)

**Comment by PagnoDunadan** [comment]
Log looks different, but still produces 403.

```
Log
Status: running collecting #27 of max 250 (keep max 25 per set, then FIFO) - change (pause to see details of the log)

Log for web-api[1]

Api.Adam[ac] - Path: https://[HOST]/[PORTAL]/DesktopModules/2sxc/API/app/auto/content/ImageLibrary/72e43bdd-d761-4b54-9e55-0bad9ce3460d/Images?folder=TestFolder/TestFolderDeep 
Api.Adam[ac]DNN.Enviro[c6]Dnn.Publsh[e6] - () 
Api.Adam[ac]DNN.Enviro[de]Dnn.Publsh[45] - () 
Api.Adam[ac]CB.Mod[fe] - parent#1518, content-block#1518, z#5, a#23 
Api.Adam[ac]CB.Mod[fe]Sxc.Instnc[08] - get SxcInstance for a:23 cb:1518 
Api.Adam[ac]CB.Mod[fe]DNN.Enviro[59]Dnn.Publsh[fa] - () 
Api.Adam[ac]CB.Mod[fe] - real app, will load data 
Api.Adam[ac]CB.Mod[fe]DNN.Enviro[71]Dnn.Publsh[5e] - () 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a] - prep App z#5, a#23, allowSE:True, hasDataConfig:True, P:3 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a] - create app resources? allowSE:True 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a]App.Zone[bf] - zone base for z#5 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a] - init app resources 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a]CG.Manage[01] - find content-group for mid#1518 and tab#426 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a]CG.Manage[01] - get CG or gen preview for grp#cdd8a08d-28e3-475c-8227-17585a001c01, preview#00000000-0000-0000-0000-000000000000 
Api.Adam[ac]CB.Mod[fe]App.2sxcAp[5a]CG.Manage[01] - get CG#cdd8a08d-28e3-475c-8227-17585a001c01 
Api.Adam[ac]CB.Mod[fe]Sxc.Instnc[08] - template override - check 
Api.Adam[ac]CB.Mod[fe]Sxc.Instnc[08] - template override - none 
Api.Adam[ac] - adam items a:23, i:72e43bdd-d761-4b54-9e55-0bad9ce3460d, field:Images, subfold:TestFolder/TestFolderDeep, useRoot:False 
Api.Adam[ac]Api.Perms[f9] - new AppAndPermissions(..., appId: 23, ...) = ready for z/a:5/23 t/z:3/5 same:True 
Api.Adam[ac]Api.Perms[f9]DNN.Enviro[4e]Dnn.Publsh[16] - () 
Api.Adam[ac]Api.Perms[f9]DNN.Enviro[9a]Dnn.Publsh[00] - () 
Api.Adam[ac]Api.Perms[f9]App.2sxcAp[6a] - prep App z#5, a#23, allowSE:False, hasDataConfig:True, P:3 
Api.Adam[ac]Api.Perms[f9]App.2sxcAp[6a] - create app resources? allowSE:False 
Api.Adam[ac]Api.Perms[f9]App.2sxcAp[6a] - init app resources 
Api.Adam[ac]Api.Perms[f9] - BuildTypePermissionChecker(ImageLibrary) 
Api.Adam[ac]Api.Perms[f9]App.Base[7d] - Zone 5, App 23 
Api.Adam[ac]Api.Perms[f9] - BuildPermissionChecker(...) 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - new PermissionCheckBase(type:83ccc1d8-1b61-42d8-913e-54103840ecb1, itm: (), permList1: 0, permList2: ) = ready 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - new DnnPermissionCheck(..., , app: 23, ) 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - UserMay([Create,Update,Approve,Full,Develop]) = False 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~EnvironmentAllows([Create,Update,Approve,Full,Develop]) = False because:Undefined 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~~~UserIsSuperuser= False 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~~~CurrentZoneMatchesTenantZone() = True 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~~~UserIsSuperuser= False 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~~~UserIsModuleAdmin= False 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~~~UserIsModuleEditor= False 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~get PermissionList start= found 1 items 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~DoesPermissionListAllow([Create, Update, Approve, Full, Develop]) for 1 permission items= False 
Api.Adam[ac]Api.Perms[f9]App.PermCk[32] - ~~~~DoesPermissionAllow(cuafv) = False 
end of log
```

**Comment by iJungleboy** [comment]
I can't figure out where this happens, so I added more logs. Could you re-run?
[ToSic.WebApi.zip](https://github.com/2sic/2sxc/files/2572629/ToSic.WebApi.zip)

**Comment by PagnoDunadan** [comment]
Here it is. Tell me if you need some other log because Insights is new for me.
```
Log
Status: running collecting #18 of max 250 (keep max 25 per set, then FIFO) - change (pause to see details of the log)

Log for web-api[1]

Api.Adam[a5] - Path: https://[HOST]/[PORTAL]/DesktopModules/2sxc/API/app/auto/content/ImageLibrary/72e43bdd-d761-4b54-9e55-0bad9ce3460d/Images?folder=TestFolder/TestFolderDeep 
Api.Adam[a5]DNN.Enviro[f7]Dnn.Publsh[81] - () 
Api.Adam[a5]DNN.Enviro[cc]Dnn.Publsh[72] - () 
Api.Adam[a5]CB.Mod[b0] - parent#1518, content-block#1518, z#5, a#23 
Api.Adam[a5]CB.Mod[b0]Sxc.Instnc[2b] - get SxcInstance for a:23 cb:1518 
Api.Adam[a5]CB.Mod[b0]DNN.Enviro[40]Dnn.Publsh[2a] - () 
Api.Adam[a5]CB.Mod[b0] - real app, will load data 
Api.Adam[a5]CB.Mod[b0]DNN.Enviro[21]Dnn.Publsh[9b] - () 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a] - prep App z#5, a#23, allowSE:True, hasDataConfig:True, P:3 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a] - create app resources? allowSE:True 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a]App.Zone[55] - zone base for z#5 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a] - init app resources 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a]CG.Manage[bc] - find content-group for mid#1518 and tab#426 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a]CG.Manage[bc] - get CG or gen preview for grp#cdd8a08d-28e3-475c-8227-17585a001c01, preview#00000000-0000-0000-0000-000000000000 
Api.Adam[a5]CB.Mod[b0]App.2sxcAp[7a]CG.Manage[bc] - get CG#cdd8a08d-28e3-475c-8227-17585a001c01 
Api.Adam[a5]CB.Mod[b0]Sxc.Instnc[2b] - template override - check 
Api.Adam[a5]CB.Mod[b0]Sxc.Instnc[2b] - template override - none 
Api.Adam[a5] - adam items a:23, i:72e43bdd-d761-4b54-9e55-0bad9ce3460d, field:Images, subfold:TestFolder/TestFolderDeep, useRoot:False
Api.Adam[a5]Api.Perms[3b] - new AppAndPermissions(..., appId: 23, ...) = ready for z/a:5/23 t/z:3/5 same:True 
Api.Adam[a5]Api.Perms[3b]DNN.Enviro[51]Dnn.Publsh[56] - () 
Api.Adam[a5]Api.Perms[3b]DNN.Enviro[08]Dnn.Publsh[fc] - () 
Api.Adam[a5]Api.Perms[3b]App.2sxcAp[0f] - prep App z#5, a#23, allowSE:False, hasDataConfig:True, P:3 
Api.Adam[a5]Api.Perms[3b]App.2sxcAp[0f] - create app resources? allowSE:False 
Api.Adam[a5]Api.Perms[3b]App.2sxcAp[0f] - init app resources 
Api.Adam[a5]Api.Perms[3b] - BuildTypePermissionChecker(ImageLibrary) 
Api.Adam[a5]Api.Perms[3b]App.Base[68] - Zone 5, App 23 
Api.Adam[a5]Api.Perms[3b] - BuildPermissionChecker(...) 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - new PermissionCheckBase(type:83ccc1d8-1b61-42d8-913e-54103840ecb1, itm: (), permList1: 0, permList2: ) = ready 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - new DnnPermissionCheck(..., , app: 23, ) 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - UserMay([Create,Update,Approve,Full,Develop]) = False 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~EnvironmentAllows([Create,Update,Approve,Full,Develop]) = False because:Undefined 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~~~UserIsSuperuser= False 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~~~CurrentZoneMatchesTenantZone() = True 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~~~UserIsSuperuser= False 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~~~UserIsModuleAdmin= False 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~~~UserIsModuleEditor= False 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~get PermissionList start= found 1 items 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~DoesPermissionListAllow([Create, Update, Approve, Full, Develop]) for 1 permission items= False 
Api.Adam[a5]Api.Perms[3b]App.PermCk[94] - ~~~~DoesPermissionAllow(cuafv) = False 
Api.Adam[a5]Api.Perms[3b] - check if feature enabled 
end of log
```

**Comment by iJungleboy** [comment]
Based on the log I believe you don't have the public-api feature enabled yet (this is a security precaution). Did you do that? https://2sxc.org/en/blog/post/new-features-management-in-2sxc-9-30

**Comment by PagnoDunadan** [comment]
I had Custom WebAPIs save in ADAM and Public Forms File Upload enabled. I couldn't find entry public-api. Closest I could find is Public forms. Enabling that produces following log.
```
Log
Status: running collecting #69 of max 250 (keep max 25 per set, then FIFO) - change (pause to see details of the log)

Log for web-api[1]

Api.Adam[4d] - Path: https://bipa.dev.sistemi.hr/bipa-backstage/DesktopModules/2sxc/API/app/auto/content/ImageLibrary/72e43bdd-d761-4b54-9e55-0bad9ce3460d/Images?folder=TestFolder/TestFolderDeep 
Api.Adam[4d]DNN.Enviro[2f]Dnn.Publsh[e8] - () 
Api.Adam[4d]DNN.Enviro[cd]Dnn.Publsh[6d] - () 
Api.Adam[4d]CB.Mod[49] - parent#1518, content-block#1518, z#5, a#23 
Api.Adam[4d]CB.Mod[49]Sxc.Instnc[3f] - get SxcInstance for a:23 cb:1518 
Api.Adam[4d]CB.Mod[49]DNN.Enviro[f9]Dnn.Publsh[fe] - () 
Api.Adam[4d]CB.Mod[49] - real app, will load data 
Api.Adam[4d]CB.Mod[49]DNN.Enviro[4a]Dnn.Publsh[70] - () 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23] - prep App z#5, a#23, allowSE:True, hasDataConfig:True, P:3 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23] - create app resources? allowSE:True 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23]App.Zone[b4] - zone base for z#5 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23] - init app resources 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23]CG.Manage[66] - find content-group for mid#1518 and tab#426 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23]CG.Manage[66] - get CG or gen preview for grp#cdd8a08d-28e3-475c-8227-17585a001c01, preview#00000000-0000-0000-0000-000000000000 
Api.Adam[4d]CB.Mod[49]App.2sxcAp[23]CG.Manage[66] - get CG#cdd8a08d-28e3-475c-8227-17585a001c01 
Api.Adam[4d]CB.Mod[49]Sxc.Instnc[3f] - template override - check 
Api.Adam[4d]CB.Mod[49]Sxc.Instnc[3f] - template override - none 
Api.Adam[4d] - adam items a:23, i:72e43bdd-d761-4b54-9e55-0bad9ce3460d, field:Images, subfold:TestFolder/TestFolderDeep, useRoot:False 
Api.Adam[4d]Api.Perms[ea] - new AppAndPermissions(..., appId: 23, ...) = ready for z/a:5/23 t/z:3/5 same:True 
Api.Adam[4d]Api.Perms[ea]DNN.Enviro[7b]Dnn.Publsh[0c] - () 
Api.Adam[4d]Api.Perms[ea]DNN.Enviro[9e]Dnn.Publsh[0b] - () 
Api.Adam[4d]Api.Perms[ea]App.2sxcAp[c7] - prep App z#5, a#23, allowSE:False, hasDataConfig:True, P:3 
Api.Adam[4d]Api.Perms[ea]App.2sxcAp[c7] - create app resources? allowSE:False 
Api.Adam[4d]Api.Perms[ea]App.2sxcAp[c7] - init app resources 
Api.Adam[4d]Api.Perms[ea] - BuildTypePermissionChecker(ImageLibrary) 
Api.Adam[4d]Api.Perms[ea]App.Base[f8] - Zone 5, App 23 
Api.Adam[4d]Api.Perms[ea] - BuildPermissionChecker(...) 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - new PermissionCheckBase(type:83ccc1d8-1b61-42d8-913e-54103840ecb1, itm: (), permList1: 0, permList2: ) = ready 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - new DnnPermissionCheck(..., , app: 23, ) 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - UserMay([Create,Update,Approve,Full,Develop]) = False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~EnvironmentAllows([Create,Update,Approve,Full,Develop]) = False because:Undefined 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~~~UserIsSuperuser= False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~~~CurrentZoneMatchesTenantZone() = True 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~~~UserIsSuperuser= False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~~~UserIsModuleAdmin= False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~~~UserIsModuleEditor= False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~get PermissionList start= found 1 items 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~DoesPermissionListAllow([Create, Update, Approve, Full, Develop]) for 1 permission items= False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[c7] - ~~~~DoesPermissionAllow(cuafv) = False 
Api.Adam[4d]Api.Perms[ea] - check if feature enabled 
Api.Adam[4d]Api.Perms[ea] - PrepCore(...) 
Api.Adam[4d]Api.Perms[ea]Adm.ApCntx[42] - starting 
Api.Adam[4d]Api.Perms[ea]App.Base[86] - Zone 5, App 23 
Api.Adam[4d]Api.Perms[ea] - FileTypeIsOkForThisField() = True 
Api.Adam[4d]Api.Perms[ea] - ~~field type:Hyperlink 
Api.Adam[4d] - starting permissions checks 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - new PermissionCheckBase(type:, itm: (), permList1: , permList2: 0) = ready 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - new DnnPermissionCheck(..., , app: , ) 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - UserMay([Approve,Read,ReadDraft,Update,UpdateDraft,Develop,Full]) = False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - ~~EnvironmentAllows([Approve,Read,ReadDraft,Update,UpdateDraft,Develop,Full]) = False because:Undefined 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - ~~~~UserIsSuperuser= False 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - ~~~~CurrentZoneMatchesTenantZone() 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - ~~~~get PermissionList start= found 0 items 
Api.Adam[4d]Api.Perms[ea]App.PermCk[d5] - ~~~~DoesPermissionListAllow([Approve, Read, ReadDraft, Update, UpdateDraft, Develop, Full]) for 0 permission items= False 
Api.Adam[4d] - user is restricted, and doesn't have permissions on field - return null 
end of log
```

**Comment by iJungleboy** [comment]
This is looking much better :).
ATM it says the user doesn't have permissions on the field. Could you verify that he has read on the content-type AND the field?

**Comment by PagnoDunadan** [comment]
Log after enabling Read rights on field. I also tried enabling CRUD on field and that works, but just Read doesn't work.
```
Log
Status: running collecting #134 of max 250 (keep max 25 per set, then FIFO) - change (pause to see details of the log)

Log for web-api[1]

Api.Adam[40] - Path: https://[HOST]/[PORTAL]/DesktopModules/2sxc/API/app/auto/content/ImageLibrary/72e43bdd-d761-4b54-9e55-0bad9ce3460d/Images?folder=TestFolder/TestFolderDeep 
Api.Adam[40]DNN.Enviro[1d]Dnn.Publsh[cf] - () 
Api.Adam[40]DNN.Enviro[d5]Dnn.Publsh[59] - () 
Api.Adam[40]CB.Mod[32] - parent#1518, content-block#1518, z#5, a#23 
Api.Adam[40]CB.Mod[32]Sxc.Instnc[ac] - get SxcInstance for a:23 cb:1518 
Api.Adam[40]CB.Mod[32]DNN.Enviro[03]Dnn.Publsh[58] - () 
Api.Adam[40]CB.Mod[32] - real app, will load data 
Api.Adam[40]CB.Mod[32]DNN.Enviro[8a]Dnn.Publsh[13] - () 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99] - prep App z#5, a#23, allowSE:True, hasDataConfig:True, P:3 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99] - create app resources? allowSE:True 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99]App.Zone[1c] - zone base for z#5 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99] - init app resources 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99]CG.Manage[c0] - find content-group for mid#1518 and tab#426 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99]CG.Manage[c0] - get CG or gen preview for grp#cdd8a08d-28e3-475c-8227-17585a001c01, preview#00000000-0000-0000-0000-000000000000 
Api.Adam[40]CB.Mod[32]App.2sxcAp[99]CG.Manage[c0] - get CG#cdd8a08d-28e3-475c-8227-17585a001c01 
Api.Adam[40]CB.Mod[32]Sxc.Instnc[ac] - template override - check 
Api.Adam[40]CB.Mod[32]Sxc.Instnc[ac] - template override - none 
Api.Adam[40] - adam items a:23, i:72e43bdd-d761-4b54-9e55-0bad9ce3460d, field:Images, subfold:TestFolder/TestFolderDeep, useRoot:False 
Api.Adam[40]Api.Perms[d3] - new AppAndPermissions(..., appId: 23, ...) = ready for z/a:5/23 t/z:3/5 same:True 
Api.Adam[40]Api.Perms[d3]DNN.Enviro[c4]Dnn.Publsh[a2] - () 
Api.Adam[40]Api.Perms[d3]DNN.Enviro[36]Dnn.Publsh[63] - () 
Api.Adam[40]Api.Perms[d3]App.2sxcAp[e8] - prep App z#5, a#23, allowSE:False, hasDataConfig:True, P:3 
Api.Adam[40]Api.Perms[d3]App.2sxcAp[e8] - create app resources? allowSE:False 
Api.Adam[40]Api.Perms[d3]App.2sxcAp[e8] - init app resources 
Api.Adam[40]Api.Perms[d3] - BuildTypePermissionChecker(ImageLibrary) 
Api.Adam[40]Api.Perms[d3]App.Base[1d] - Zone 5, App 23 
Api.Adam[40]Api.Perms[d3] - BuildPermissionChecker(...) 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - new PermissionCheckBase(type:83ccc1d8-1b61-42d8-913e-54103840ecb1, itm: (), permList1: 0, permList2: ) = ready 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - new DnnPermissionCheck(..., , app: 23, ) 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - UserMay([Create,Update,Approve,Full,Develop]) = False 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~EnvironmentAllows([Create,Update,Approve,Full,Develop]) = False because:Undefined 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~~~UserIsSuperuser= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~~~CurrentZoneMatchesTenantZone() = True 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~~~UserIsSuperuser= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~~~UserIsModuleAdmin= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~~~UserIsModuleEditor= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~get PermissionList start= found 1 items 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~DoesPermissionListAllow([Create, Update, Approve, Full, Develop]) for 1 permission items= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[ff] - ~~~~DoesPermissionAllow(cuafv) = False 
Api.Adam[40]Api.Perms[d3] - check if feature enabled 
Api.Adam[40]Api.Perms[d3] - PrepCore(...) 
Api.Adam[40]Api.Perms[d3]Adm.ApCntx[de] - starting 
Api.Adam[40]Api.Perms[d3]App.Base[51] - Zone 5, App 23 
Api.Adam[40]Api.Perms[d3] - FileTypeIsOkForThisField() = True 
Api.Adam[40]Api.Perms[d3] - ~~field type:Hyperlink 
Api.Adam[40] - starting permissions checks 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - new PermissionCheckBase(type:, itm: (), permList1: , permList2: 1) = ready 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - new DnnPermissionCheck(..., , app: , ) 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - UserMay([Approve,Read,ReadDraft,Update,UpdateDraft,Develop,Full]) = True 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - ~~EnvironmentAllows([Approve,Read,ReadDraft,Update,UpdateDraft,Develop,Full]) = False because:Undefined 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - ~~~~UserIsSuperuser= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - ~~~~CurrentZoneMatchesTenantZone() 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - ~~~~get PermissionList start= found 1 items 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - ~~~~DoesPermissionListAllow([Approve, Read, ReadDraft, Update, UpdateDraft, Develop, Full]) for 1 permission items= True 
Api.Adam[40]Api.Perms[d3]App.PermCk[7f] - ~~~~~~DoesPermissionAllow(arřuǔvf) = True 
Api.Adam[40]Api.Perms[d3] - check if user is restricted (True) or if the item '72e43bdd-d761-4b54-9e55-0bad9ce3460d' is draft 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - new PermissionCheckBase(type:, itm: (), permList1: , permList2: 1) = ready 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - new DnnPermissionCheck(..., , app: , ) 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - UserMay([Approve,Delete,Develop,Full,Update]) = False 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - ~~EnvironmentAllows([Approve,Delete,Develop,Full,Update]) = False because:Undefined 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - ~~~~UserIsSuperuser= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - ~~~~CurrentZoneMatchesTenantZone() 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - ~~~~get PermissionList start= found 1 items 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - ~~~~DoesPermissionListAllow([Approve, Delete, Develop, Full, Update]) for 1 permission items= False 
Api.Adam[40]Api.Perms[d3]App.PermCk[23] - ~~~~~~DoesPermissionAllow(advfu) = False 
Api.Adam[40]Api.Perms[d3] - user is restricted and may not see published, but item exists and is published - not allowed 
Api.Adam[40] - user is restricted and item is already published 
end of log
```

**Comment by iJungleboy** [comment]
This looks good - I believe a permission list (is user allowed to read published) wasn't complete. I just added this - probably all we need. Check out the new DLLs.
[ToSic.WebApi.zip](https://github.com/2sic/2sxc/files/2576855/ToSic.WebApi.zip)

**Comment by PagnoDunadan** [comment]
Yes! It works :)

**Comment by iJungleboy** [comment]
Awesome. Will be in 9.34 👍

**Comment by tpperlman** [comment]
Hi folks! I have tried following the instructions below but am unable to access ADAM assets through the WebAPI.

I have posted a thread on StackOverflow at [https://stackoverflow.com/questions/55768099/webapi-error-accessing-adam-assets-in-2sxc-v9-42](https://stackoverflow.com/questions/55768099/webapi-error-accessing-adam-assets-in-2sxc-v9-42) to request assistance. Perhaps someone can provide some guidance.
